# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю

Выполнили студенты гр. 2381 Ивашинников Леонид Дмитриевич и Двиков Даниил Владиславович. Вариант №23.

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Краткое изложение основных теоретических понятий

Корреляционная зависимость
Корреляционной называют статистическую зависимость двух случайных величин, при которой изменение значения одной случайной величины приводит к изменению математического ожидания другой. Функции $M(X \mid y)$ и $M(Y \mid x)$ называются функциями регрессии.

Корреляционный момент определяется формулой
$
\mu_{xy}=M\big[(X-M(X))(Y-M(Y))\big].
$

Коэффициент корреляции вычисляется по формуле
$
r_{xy}=\frac{\mu_{xy}}{\sigma_x \sigma_y}.
$

Он характеризует тесноту линейной связи между случайными величинами $X$ и $Y$.

Свойства коэффициента корреляции
Для коэффициента корреляции выполняется условие
$
|r_{xy}| \le 1.
$

Если коэффициент корреляции отличен от нуля, то случайные величины являются коррелированными. Если величины коррелированы, то они зависимы. Однако из некоррелированности не следует независимость. При $|r_{xy}|=1$ линейная зависимость становится функциональной.

Выборочный коэффициент корреляции
Статистическая оценка коэффициента корреляции вычисляется по стандартной формуле
$
\bar r_{xy}=
\frac{\sum\sum n_{ij}x_i y_j - N\bar x \bar y}
{N S_x S_y}.
$

Эта формула применяется при работе с корреляционной таблицей и позволяет оценить тесноту линейной связи между исследуемыми признаками.

Вычисление коэффициента корреляции с помощью условных вариант
Для упрощения вычислений используются условные варианты:
$
u_i=\frac{x_i-C_x}{h_x}, \qquad v_j=\frac{y_j-C_y}{h_y}.
$

Тогда выборочный коэффициент корреляции вычисляют через условные варианты.
 
$
r_{в}=
\frac{m_{11}-m_{10}m_{01}}
{\sqrt{m_{20}-m_{10}^2}\sqrt{m_{02}-m_{01}^2}}
$

Доверительный интервал для коэффициента корреляции
Для построения доверительного интервала используется преобразование Фишера:
$
z=0.5\ln\frac{1+\bar r_{xy}}{1-\bar r_{xy}}.
$

При больших объёмах выборки величина $z$ имеет асимптотически нормальное распределение со среднеквадратическим отклонением
$
\sigma_z=\frac{1}{\sqrt{N-3}}.
$

Сначала строят доверительный интервал для $z$, а затем переходят к доверительному интервалу для коэффициента корреляции с помощью обратного преобразования Фишера:
$
r=\operatorname{th}(z).
$

Проверка гипотезы о значимости коэффициента корреляции
Проверяется нулевая гипотеза
$
H_0: r_{xy}=0,
$

и альтернативная гипотеза
$
H_1: r_{xy}\ne 0.
$

В качестве критерия используется статистика
$
T=\frac{\bar r_{xy}\sqrt{N-2}}{\sqrt{1-\bar r_{xy}^2}}.
$

При справедливости нулевой гипотезы эта статистика имеет распределение Стьюдента с числом степеней свободы $k=N-2$. Если
$
|T_{\text{набл}}| > t_{\text{крит}}(\alpha, k),
$
то гипотеза $H_0$ отвергается; иначе оснований для её отклонения нет.

## Постановка задачи

Из заданной генеральной совокупности необходимо сформировать выборку по второму признаку и провести её статистическую обработку в объёме практических работ №1 и №2. Требуется определить точечные статистические оценки параметров распределения исследуемого признака: математическое ожидание, дисперсию, среднеквадратичное отклонение, асимметрию, эксцесс и коэффициент вариации.

Для системы двух случайных величин $X$ (первый признак) и $Y$ (второй признак) необходимо сформировать двумерную выборку, найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и выполнить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты требуется содержательно интерпретировать.

## Выполнение работы

### 1. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2 с целью определения точечных статистических оценок параметров распределения исследуемого признака: математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации. Результаты оформить в виде таблицы и сделать выводы.


In [ ]:
import pandas as pd


NU_HEADER, E_HEADER = 'nu (kg/cm^3)', 'E (kg/cm^2)'

df = pd.read_csv('/content/dataset.csv', sep=';')
df.describe()

,nu (kg/cm^3),E (kg/cm^2)
count,109.000000,109.000000
mean,0.449569,127.912844
std,0.054279,22.346406
min,0.320000,71.100000
25%,0.412000,114.800000
50%,0.448000,127.700000
75%,0.487000,143.800000
max,0.593000,187.400000


In [123]:
import numpy as np
import pandas as pd
import math

INTERVAL = r'$[x_i,x_{i+1})$'
MIDRANGE = r'$\tilde x_i$'
FREQUENCY = r'$m_i$'
RELATIVE_FREQUENCY = r'$\tilde m_i$'
FREQUENCY_CUM_SUM = r'$m_i^{нак}$'
RELATIVE_FREQUENCY_CUM_SUM = r'$\tilde m_i^{нак}$'
SUM = r'$\sum$'


def sturges_interval_series(s: pd.Series, precision: int = 4) -> pd.DataFrame:
    n = len(s)
    k = math.ceil(1 + 3.322 * np.log10(n))

    xmin, xmax = float(s.min()), float(s.max())
    h = (xmax - xmin) / k

    bins = [round(xmin + i * h, precision) for i in range(k)] + [round(xmax, precision)]
    bins[-1] += 1e-12

    intervals = pd.cut(s, bins=bins, right=False, include_lowest=True)

    table = intervals.value_counts().sort_index().reset_index()

    table.columns = [INTERVAL, FREQUENCY]
    table[MIDRANGE] = table[INTERVAL].apply(lambda x: round((x.left + x.right) / 2, precision))
    table[RELATIVE_FREQUENCY] = round(table[FREQUENCY] / n, precision)
    table[FREQUENCY_CUM_SUM] = table[FREQUENCY].cumsum()
    table[RELATIVE_FREQUENCY_CUM_SUM] = round(table[RELATIVE_FREQUENCY].cumsum(), precision)

    table.index = range(1, len(table) + 1)

    sum_row = pd.DataFrame({
        INTERVAL: [''],
        MIDRANGE: [''],
        FREQUENCY: [table[FREQUENCY].sum()],
        RELATIVE_FREQUENCY: [round(table[RELATIVE_FREQUENCY].sum(), precision)],
        FREQUENCY_CUM_SUM: [''],
        RELATIVE_FREQUENCY_CUM_SUM: ['']
    }, index=[SUM])

    table = pd.concat([table, sum_row])

    return table[
        [
            INTERVAL,
            MIDRANGE,
            FREQUENCY,
            RELATIVE_FREQUENCY,
            FREQUENCY_CUM_SUM,
            RELATIVE_FREQUENCY_CUM_SUM
        ]
    ]
interval_e = sturges_interval_series(df[E_HEADER])

In [124]:
from IPython.display import display, Markdown


display(Markdown(interval_e.to_markdown()))

|        | $[x_i,x_{i+1})$    | $\tilde x_i$   |   $m_i$ |   $\tilde m_i$ | $m_i^{нак}$   | $\tilde m_i^{нак}$   |
|:-------|:-------------------|:---------------|--------:|---------------:|:--------------|:---------------------|
| 1      | [71.1, 85.638)     | 78.369         |       7 |         0.0642 | 7             | 0.0642               |
| 2      | [85.638, 100.175)  | 92.9065        |       5 |         0.0459 | 12            | 0.1101               |
| 3      | [100.175, 114.712) | 107.4435       |      15 |         0.1376 | 27            | 0.2477               |
| 4      | [114.712, 129.25)  | 121.981        |      30 |         0.2752 | 57            | 0.5229               |
| 5      | [129.25, 143.788)  | 136.519        |      24 |         0.2202 | 81            | 0.7431               |
| 6      | [143.788, 158.325) | 151.0565       |      19 |         0.1743 | 100           | 0.9174               |
| 7      | [158.325, 172.862) | 165.5935       |       7 |         0.0642 | 107           | 0.9816               |
| 8      | [172.862, 187.4)   | 180.131        |       2 |         0.0183 | 109           | 0.9999               |
| $\sum$ |                    |                |     109 |         0.9999 |               |                      |

In [184]:
import numpy as np
import pandas as pd


X = r'$x_i$'
N = r'$n_i$'
U = r'$u_i$'
NU = r'$n_iu_i$'
NU2 = r'$n_iu_i^2$'
NU3 = r'$n_iu_i^3$'
NU4 = r'$n_iu_i^4$'
NUPLUS14 = r'$n_i(u_i+1)^4$'


def build_variant(table: pd.DataFrame) -> pd.DataFrame:
    midranges = table[MIDRANGE][:-1].astype(float)  # 1. привести к float

    idx_modal = table[FREQUENCY][:-1].astype(int).values.argmax()
    c = midranges.iloc[idx_modal]
    h = table[INTERVAL].iloc[0].length               

    u = np.round((midranges - c) / h).astype(int)
    nu = table[FREQUENCY][:-1] * u
    nu2 = table[FREQUENCY][:-1] * (u ** 2)
    nu3 = table[FREQUENCY][:-1] * (u ** 3)
    nu4 = table[FREQUENCY][:-1] * (u ** 4)
    nuplus14 = table[FREQUENCY][:-1] * ((u + 1) ** 4)

    result = pd.DataFrame({
        X: midranges,
        N: table[FREQUENCY][:-1],
        U: u,
        NU: nu,
        NU2: nu2,
        NU3: nu3,
        NU4: nu4,
        NUPLUS14: nuplus14
    })

    sum_row = pd.DataFrame({
        X: [''],
        N: [result[N].sum()],
        U: [''],
        NU: [result[NU].sum()],
        NU2: [result[NU2].sum()],
        NU3: [result[NU3].sum()],
        NU4: [result[NU4].sum()],
        NUPLUS14: [result[NUPLUS14].sum()],
    }, index=[SUM])

    return pd.concat([result, sum_row]).rename_axis('i')


variant_e =  build_variant(interval_e)

In [142]:
from IPython.display import display, Markdown


display(Markdown(variant_e.to_markdown()))

| i      | $x_i$    |   $n_i$ | $u_i$   |   $n_iu_i$ |   $n_iu_i^2$ |   $n_iu_i^3$ |   $n_iu_i^4$ |   $n_i(u_i+1)^4$ |
|:-------|:---------|--------:|:--------|-----------:|-------------:|-------------:|-------------:|-----------------:|
| 1      | 78.369   |       7 | -3      |        -21 |           63 |         -189 |          567 |              112 |
| 2      | 92.9065  |       5 | -2      |        -10 |           20 |          -40 |           80 |                5 |
| 3      | 107.4435 |      15 | -1      |        -15 |           15 |          -15 |           15 |                0 |
| 4      | 121.981  |      30 | 0       |          0 |            0 |            0 |            0 |               30 |
| 5      | 136.519  |      24 | 1       |         24 |           24 |           24 |           24 |              384 |
| 6      | 151.0565 |      19 | 2       |         38 |           76 |          152 |          304 |             1539 |
| 7      | 165.5935 |       7 | 3       |         21 |           63 |          189 |          567 |             1792 |
| 8      | 180.131  |       2 | 4       |          8 |           32 |          128 |          512 |             1250 |
| $\sum$ |          |     109 |         |         45 |          293 |          249 |         2069 |             5112 |

In [145]:
# Проверка вычислений
left = round(variant_e[NUPLUS14][SUM], 6)
right = round(
    variant_e[NU4][SUM]
    + 4 * variant_e[NU3][SUM]
    + 6 * variant_e[NU2][SUM]
    + 4 * variant_e[NU][SUM]
    + variant_e[N][SUM],
    6
)

print(f"Σ n_i(u_i+1)^4 = {left}")
print(f"Σ n_i·u^4 + 4·Σn_i·u^3 + 6·Σn_i·u^2 + 4·Σn_i·u + N = {right}")

if left == right:
    print("Вычисления верны!")
else:
    print("Вычисления ошибочны!")

Σ n_i(u_i+1)^4 = 5112
Σ n_i·u^4 + 4·Σn_i·u^3 + 6·Σn_i·u^2 + 4·Σn_i·u + N = 5112
Вычисления верны!


In [148]:
MOMENT = r'$\bar M_k^*$'
CENTRAL_MOMENT = r'$\bar m_k$'


def build_moments(table: pd.DataFrame, h: float) -> pd.DataFrame:
  m = table[[NU, NU2, NU3, NU4]].loc[SUM] / table[N].loc[SUM]
  central_m = [
    0,
    (m[NU2]-m[NU]**2) * (h**2),
    (m[NU3]-3*m[NU2]*m[NU]+2*(m[NU]**3)) * (h**3),
    (m[NU4]-4*m[NU3]*m[NU]+6*m[NU2]*(m[NU]**2)-3*(m[NU]**4)) * (h**4)
  ]

  result = pd.DataFrame({
      MOMENT: m.to_list(),
      CENTRAL_MOMENT: central_m
  }, index=list(range(1, len(m)+1)))

  return result.rename_axis('k')


moment_e = build_moments(variant_e, interval_e[INTERVAL][1].length)


In [147]:
from IPython.display import display, Markdown


display(Markdown(moment_e.to_markdown()))

|   k |   $\bar M_k^*$ |   $\bar m_k$ |
|----:|---------------:|-------------:|
|   1 |       0.412844 |         0    |
|   2 |       2.68807  |       532.11 |
|   3 |       2.2844   |     -2778.08 |
|   4 |      18.9817   |    798304    |

In [186]:
# Извлекаем суммы из таблицы variant_e
N_val = variant_e[N][SUM]
S1 = variant_e[NU][SUM]
S2 = variant_e[NU2][SUM]
S3 = variant_e[NU3][SUM]
S4 = variant_e[NU4][SUM]

# Параметры из интервального ряда
midranges = interval_e[MIDRANGE][:-1].astype(float)
frequencies = interval_e[FREQUENCY][:-1].astype(int)
h = interval_e[INTERVAL].iloc[0].length

idx_modal = frequencies.values.argmax()
C = midranges.iloc[idx_modal]

# Условные моменты
M1 = S1 / N_val
M2 = S2 / N_val
M3 = S3 / N_val
M4 = S4 / N_val

#Центральные моменты
m2 = (M2 - M1**2) * h**2
m3 = (M3 - 3*M2*M1 + 2*M1**3) * h**3
m4 = (M4 - 4*M3*M1 + 6*M2*M1**2 - 3*M1**4) * h**4

x_mean = C + h * M1

D = m2

sigma = np.sqrt(D)

S2_corr = (N_val / (N_val - 1)) * D
S_corr = np.sqrt(S2_corr)

A_s = m3 / S_corr**3

E_k = m4 / S_corr**4 - 3

# Мода
n_mo = frequencies.iloc[idx_modal]
n_mo_prev = frequencies.iloc[idx_modal - 1] if idx_modal > 0 else 0
n_mo_next = frequencies.iloc[idx_modal + 1] if idx_modal < len(frequencies) - 1 else 0
x_mo = midranges.iloc[idx_modal] - h / 2

Mo = x_mo + h * (n_mo - n_mo_prev) / (2 * n_mo - n_mo_prev - n_mo_next)

# Медиана
cum_freq = frequencies.cumsum()
me_idx = (cum_freq >= N_val / 2).argmax()
x_me = midranges.iloc[me_idx] - h / 2
n_me = frequencies.iloc[me_idx]
S_me_prev = cum_freq.iloc[me_idx - 1] if me_idx > 0 else 0

Me = x_me + h * (N_val / 2 - S_me_prev) / n_me

# Коэффициент вариации
V = (sigma / x_mean) * 100


stats_df = pd.DataFrame({
    'Параметр': [
        r'$\bar{x}$',
        r'$D$',
        r'$S^2$ (испр.)',
        r'$\sigma$',
        r'$S$ (испр.)',
        r'$A_s$',
        r'$E_k$',
        r'$M_o$',
        r'$M_e$',
        r'$V$ (%)'
    ],
    'Значение': [
        round(x_mean, 4),
        round(D, 4),
        round(S2_corr, 4),
        round(sigma, 4),
        round(S_corr, 4),
        round(A_s, 4),
        round(E_k, 4),
        round(Mo, 4),
        round(Me, 4),
        round(V, 2)
    ]
})
stats_df.index = range(1, len(stats_df) + 1)
display(Markdown(stats_df.to_markdown()))

|    | Параметр      |   Значение |
|---:|:--------------|-----------:|
|  1 | $\bar{x}$     |   127.983  |
|  2 | $D$           |   532.11   |
|  3 | $S^2$ (испр.) |   537.037  |
|  4 | $\sigma$      |    23.0675 |
|  5 | $S$ (испр.)   |    23.1741 |
|  6 | $A_s$         |    -0.2232 |
|  7 | $E_k$         |    -0.232  |
|  8 | $M_o$         |   125.096  |
|  9 | $M_e$         |   128.038  |
| 10 | $V$ (%)       |    18.02   |

Среднее значение признака составляет $\bar{x} = 127.9829$. Коэффициент вариации $V = 18.02\%$, что соответствует среднему уровню изменчивости признака.

Асимметрия $A_s = -0.2232 < 0$ указывает на слабую левостороннюю асимметрию — хвост распределения незначительно вытянут влево. Эксцесс $E_k = -0.2320 < 0$ означает, что распределение более плосковершинное (платикуртическое) по сравнению с нормальным.

Мода $M_o = 125.0963$, медиана $M_e = 128.0385$, среднее $\bar{x} = 127.9829$. Близость этих трёх величин друг к другу свидетельствует о том, что распределение признака близко к нормальному. Соотношение $M_o < M_e \approx \bar{x}$ согласуется с наличием слабой левосторонней асимметрии.

### 2. Построить двумерный интервальный вариационный ряд, оформить в виде таблицы.

In [ ]:
# Интервалы для X (E) и Y (NU) 
n = len(df)
k_e = math.ceil(1 + 3.322 * np.log10(n))
xmin_e, xmax_e = float(df[E_HEADER].min()), float(df[E_HEADER].max())
h_e = (xmax_e - xmin_e) / k_e
bins_e = [xmin_e + i * h_e for i in range(k_e)] + [xmax_e]
bins_e[-1] += 1e-12

# Построение bins для NU
k_nu = math.ceil(1 + 3.322 * np.log10(n))
xmin_nu, xmax_nu = float(df[NU_HEADER].min()), float(df[NU_HEADER].max())
h_nu = (xmax_nu - xmin_nu) / k_nu
bins_nu = [xmin_nu + i * h_nu for i in range(k_nu)] + [xmax_nu]
bins_nu[-1] += 1e-12

# Разбиение
x_intervals = pd.cut(df[E_HEADER], bins=bins_e, right=False, include_lowest=True)
y_intervals = pd.cut(df[NU_HEADER], bins=bins_nu, right=False, include_lowest=True)

# Корреляционная таблица 
corr_table = pd.crosstab(y_intervals, x_intervals, margins=True, margins_name='n')

corr_table.index.name = NU_HEADER
corr_table.columns.name = E_HEADER

display(Markdown(corr_table.to_markdown()))

| nu (kg/cm^3)   |   [71.1, 85.637) |   [85.637, 100.175) |   [100.175, 114.712) |   [114.712, 129.25) |   [129.25, 143.788) |   [143.788, 158.325) |   [158.325, 172.862) |   [172.862, 187.4) |   n |
|:---------------|-----------------:|--------------------:|---------------------:|--------------------:|--------------------:|---------------------:|---------------------:|-------------------:|----:|
| [0.32, 0.354)  |                4 |                   1 |                    1 |                   0 |                   0 |                    0 |                    0 |                  0 |   6 |
| [0.354, 0.388) |                1 |                   3 |                    3 |                   0 |                   0 |                    0 |                    0 |                  0 |   7 |
| [0.388, 0.422) |                2 |                   1 |                   10 |                   9 |                   1 |                    0 |                    0 |                  0 |  23 |
| [0.422, 0.456) |                0 |                   0 |                    0 |                  16 |                   7 |                    0 |                    0 |                  0 |  23 |
| [0.456, 0.491) |                0 |                   0 |                    1 |                   5 |                  13 |                    5 |                    0 |                  0 |  24 |
| [0.491, 0.525) |                0 |                   0 |                    0 |                   0 |                   3 |                    8 |                    3 |                  1 |  15 |
| [0.525, 0.559) |                0 |                   0 |                    0 |                   0 |                   0 |                    6 |                    4 |                  0 |  10 |
| [0.559, 0.593) |                0 |                   0 |                    0 |                   0 |                   0 |                    0 |                    0 |                  1 |   1 |
| n              |                7 |                   5 |                   15 |                  30 |                  24 |                   19 |                    7 |                  2 | 109 |

### 3. По полученному двумерному интервальному вариационному ряду построить корреляционную таблицу, сделать выводы.

In [188]:
# Матрица частот n_ij
n_ij = corr_table.iloc[:-1, :-1].values
N_val = n_ij.sum()

# Середины интервалов
mid_x = np.array([(bins_e[i] + bins_e[i+1]) / 2 for i in range(k_e)])
mid_y = np.array([(bins_nu[i] + bins_nu[i+1]) / 2 for i in range(k_nu)])

X_i = np.array([sum(n_ij[i, j] * mid_x[j] for j in range(k_e)) for i in range(k_nu)])

# y_i * X_i
yX_i = mid_y * X_i

Y_j = np.array([sum(n_ij[i, j] * mid_y[i] for i in range(k_nu)) for j in range(k_e)])

xY_j = mid_x * Y_j

col_labels = [f"[{bins_e[j]:.4f}, {bins_e[j+1]:.4f})" for j in range(k_e)]
row_labels = [f"[{bins_nu[i]:.4f}, {bins_nu[i+1]:.4f})" for i in range(k_nu)]

corr_df = pd.DataFrame(n_ij, columns=col_labels, index=row_labels)
corr_df[r'$X_i$'] = np.round(X_i, 4)
corr_df[r'$y_i X_i$'] = np.round(yX_i, 4)

yj_row = pd.Series(list(np.round(Y_j, 4)) + [np.round(Y_j.sum(), 4), ''], index=corr_df.columns, name=r'$Y_j$')

xyj_row = pd.Series(list(np.round(xY_j, 4)) + ['', np.round(yX_i.sum(), 4)], index=corr_df.columns, name=r'$x_j Y_j$')

corr_full = pd.concat([corr_df, pd.DataFrame([yj_row, xyj_row])])

display(Markdown(corr_full.to_markdown()))

# Проверка двумя способами
print(f"Σ y_i * X_i = {yX_i.sum():.4f}")
print(f"Σ x_j * Y_j = {xY_j.sum():.4f}")

|                  |   [71.1000, 85.6375) |   [85.6375, 100.1750) |   [100.1750, 114.7125) |   [114.7125, 129.2500) |   [129.2500, 143.7875) |   [143.7875, 158.3250) |   [158.3250, 172.8625) |   [172.8625, 187.4000) | $X_i$     | $y_i X_i$   |
|:-----------------|---------------------:|----------------------:|-----------------------:|-----------------------:|-----------------------:|-----------------------:|-----------------------:|-----------------------:|:----------|:------------|
| [0.3200, 0.3541) |               4      |                1      |                 1      |                 0      |                 0      |                 0      |                 0      |                 0      | 513.825   | 173.1911    |
| [0.3541, 0.3882) |               1      |                3      |                 3      |                 0      |                 0      |                 0      |                 0      |                 0      | 679.4187  | 252.1917    |
| [0.3882, 0.4224) |               2      |                1      |                10      |                 9      |                 1      |                 0      |                 0      |                 0      | 2558.4312 | 1036.9642   |
| [0.4224, 0.4565) |               0      |                0      |                 0      |                16      |                 7      |                 0      |                 0      |                 0      | 2907.3312 | 1277.5904   |
| [0.4565, 0.4906) |               0      |                0      |                 1      |                 5      |                13      |                 5      |                 0      |                 0      | 3247.375  | 1537.835    |
| [0.4906, 0.5248) |               0      |                0      |                 0      |                 0      |                 3      |                 8      |                 3      |                 1      | 2294.9188 | 1165.1016   |
| [0.5248, 0.5589) |               0      |                0      |                 0      |                 0      |                 0      |                 6      |                 4      |                 0      | 1568.7125 | 849.948     |
| [0.5589, 0.5930) |               0      |                0      |                 0      |                 0      |                 0      |                 0      |                 0      |                 1      | 180.1313  | 103.7443    |
| $Y_j$            |               2.5301 |                1.8559 |                 5.9773 |                13.0466 |                11.1608 |                 9.6802 |                 3.6903 |                 1.0836 | 49.0248   |             |
| $x_j Y_j$        |             198.278  |              172.428  |               642.225  |              1591.44   |              1523.65   |              1462.25   |               611.093  |               195.195  |           | 6396.5664   |

Σ y_i * X_i = 6396.5664
Σ x_j * Y_j = 6396.5664


Частоты $n_{ij}$ сосредоточены вдоль главной диагонали таблицы: малым значениям $X$ соответствуют малые значения $Y$, а большим — большие. Это указывает на наличие прямой (положительной) корреляционной связи между признаками.


### 4. Исходя из результатов корреляционной таблицы вычислить значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант. Убедиться, что результаты совпадают. Сделать выводы.

In [175]:
# стандартная формула 
freq = corr_table.iloc[:-1, :-1].to_numpy(dtype=float)
n = freq.sum()
n_x, n_y = freq.sum(axis=0), freq.sum(axis=1)

x_bar = (n_x * mid_x).sum() / n
y_bar = (n_y * mid_y).sum() / n
xy_bar = (freq * np.outer(mid_y, mid_x)).sum() / n
D_x = (n_x * mid_x**2).sum() / n - x_bar**2
D_y = (n_y * mid_y**2).sum() / n - y_bar**2
cov_xy = xy_bar - x_bar * y_bar
r_standard = cov_xy / np.sqrt(D_x * D_y)

print("Способ 1: стандартная формула")
print(f"  x_bar = {x_bar:.6f},  y_bar = {y_bar:.6f}")
print(f"  D_x = {D_x:.6f},  D_y = {D_y:.6f}")
print(f"  cov_xy = {cov_xy:.6f}")
print(f"  r = {r_standard:.6f}")

# условные варианты
C_x, C_y = mid_x[len(mid_x) // 2], mid_y[len(mid_y) // 2]
u, v = (mid_x - C_x) / h_e, (mid_y - C_y) / h_nu

u_bar = (n_x * u).sum() / n
v_bar = (n_y * v).sum() / n
uv_bar = (freq * np.outer(v, u)).sum() / n
sigma_u = np.sqrt((n_x * u**2).sum() / n - u_bar**2)
sigma_v = np.sqrt((n_y * v**2).sum() / n - v_bar**2)
r_cond = (uv_bar - u_bar * v_bar) / (sigma_u * sigma_v)

print(f"\nСпособ 2: условные варианты")
print(f"  C_x = {C_x},  h_x = {h_e:.6f}")
print(f"  C_y = {C_y},  h_y = {h_nu:.6f}")
print(f"  u_bar = {u_bar:.6f},  v_bar = {v_bar:.6f}")
print(f"  r = {r_cond:.6f}")

# по исходным данным
r_raw = np.corrcoef(df[E_HEADER], df[NU_HEADER])[0, 1]
print(f"\nСпособ 3: по исходным данным")
print(f"  r = {r_raw:.6f}")

#Проверка
print(f"\nРазница |r1 - r2| = {abs(r_standard - r_cond):.2e}")

r_table = pd.DataFrame({
    'Способ': ['Стандартная формула', 'Условные варианты', 'По исходным негруппированным данным'],
    'r_xy': [round(r_standard, 6), round(r_cond, 6), round(r_raw, 6)]
})
r_table.index = range(1, len(r_table) + 1)
r_table

Способ 1: стандартная формула
  x_bar = 127.982970,  y_bar = 0.449769
  D_x = 532.073846,  D_y = 0.003045
  cov_xy = 1.121333
  r = 0.880967

Способ 2: условные варианты
  C_x = 136.51875,  h_x = 14.537500
  C_y = 0.4735625,  h_y = 0.034125
  u_bar = -0.587156,  v_bar = -0.697248
  r = 0.880967

Способ 3: по исходным данным
  r = 0.910061

Разница |r1 - r2| = 4.77e-15


,Способ,r_xy
1,Стандартная формула,0.880967
2,Условные варианты,0.880967
3,По исходным негруппированным данным,0.910061


Выборочный коэффициент корреляции, вычисленный стандартной формулой и через условные варианты, совпадает: $\bar{r}_{xy} = 0.8810$. Разница между способами составляет $4.77 \cdot 10^{-15}$, что подтверждает корректность вычислений.

Значение по негруппированным данным $r = 0.9101$ несколько выше из-за потери точности при группировке в интервалы.

Значение $|\bar{r}_{xy}| \approx 0.88$ указывает на сильную прямую линейную связь между признаками $X$ и $Y$.

### 5. Построить доверительные интервалы для коэффициента корреляции при $\gamma \in \{0.95, 0.99\}$ с помощью преобразования Фишера.

In [179]:
from scipy.stats import norm

r_xy = r_standard
N = int(n)

z_bar = 0.5 * np.log((1 + r_xy) / (1 - r_xy))
sigma_z = 1 / np.sqrt(N - 3)

rows = []
for gamma in [0.95, 0.99]:
    lam = norm.ppf(0.5 + gamma / 2)
    z_low = z_bar - lam * sigma_z
    z_high = z_bar + lam * sigma_z
    r_low = np.tanh(z_low)
    r_high = np.tanh(z_high)
    rows.append([gamma, round(lam, 4), round(r_low, 4), round(r_high, 4)])

conf_table = pd.DataFrame(rows, columns=[r'γ', 'u', 'Нижняя граница', 'Верхняя граница'])
conf_table

,γ,u,Нижняя граница,Верхняя граница
0,0.95,1.9600,0.8305,0.9171
1,0.99,2.5758,0.8110,0.9261


Доверительные интервалы для генерального коэффициента корреляции построены с помощью преобразования Фишера. При $\gamma = 0.95$: $r_{xy} \in (0.8305;\; 0.9171)$, при $\gamma = 0.99$: $r_{xy} \in (0.8110;\; 0.9261)$. В обоих случаях интервал не содержит нуля и целиком лежит в области сильной положительной корреляции, что подтверждает значимость линейной связи между $E$ и $\nu$.

### 6. Осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю при заданном уровне значимости $\alpha = 0.05$, сделать выводы

Проверим гипотезу $H_0: r_{xy} = 0$ при уровне значимости $\alpha = 0.05$. Согласно лекции (слайд 14), в качестве критерия используется величина $T = \frac{\bar{r}_{xy} \sqrt{N-2}}{\sqrt{1 - \bar{r}_{xy}^2}}$, которая при справедливости $H_0$ распределена по закону Стьюдента с $k = N - 2$ степенями свободы. Критическая область двусторонняя: если $|T_{\text{набл}}| > t_{\text{крит}}(\alpha, k)$, гипотеза $H_0$ отвергается.

In [183]:
from scipy.stats import t

alpha = 0.05
r_xy = r_standard
N = int(n)
k = N - 2

T_obs = r_xy * np.sqrt(N - 2) / np.sqrt(1 - r_xy**2)
t_crit = t.ppf(1 - alpha / 2, k)

result = "Отвергается" if abs(T_obs) > t_crit else "Не отвергается"

hyp_table = pd.DataFrame({
    'Параметр': [
        'alpha',
        'N',
        'k = N - 2',
        'r_xy',
        'T_набл',
        't_крит',
        '|T_набл| > t_крит',
        'Решение по H0'
    ],
    'Значение': [
        alpha,
        N,
        k,
        round(r_xy, 6),
        round(T_obs, 4),
        round(t_crit, 4),
        f'{abs(T_obs):.4f} > {t_crit:.4f}' if abs(T_obs) > t_crit else f'{abs(T_obs):.4f} ≤ {t_crit:.4f}',
        result
    ]
})
hyp_table

,Параметр,Значение
0,alpha,0.05
1,N,109
2,k = N - 2,107
3,r_xy,0.880967
4,T_набл,19.2587
5,t_крит,1.9824
6,|T_набл| > t_крит,19.2587 > 1.9824
7,Решение по H0,Отвергается


Поскольку $|T_{\text{набл}}| = 19.2587 > t_{\text{крит}} = 1.9824$, гипотеза $H_0: r_{xy} = 0$ отвергается при $\alpha = 0.05$. Коэффициент корреляции $\bar{r}_{xy} = 0.8810$ значимо отличается от нуля, что подтверждает наличие статистически значимой линейной связи между признаками $X$ и $Y$.

## Выводы

В ходе практической работы выполнен корреляционный анализ системы двух случайных величин $X$ и $Y$ по выборке объёмом $N = 109$.

1. Проведена статистическая обработка второго признака ($X$). Получены точечные оценки: $\bar{x} = 127.98$, $\sigma = 23.07$, $A_s = -0.22$, $E_k = -0.23$, $M_o = 125.10$, $M_e = 128.04$, $V = 18.02\%$. Близость моды, медианы и среднего, а также малые значения асимметрии и эксцесса указывают на распределение, близкое к нормальному.

2. Построен двумерный интервальный ряд и корреляционная таблица. Частоты $n_{ij}$ сосредоточены вдоль главной диагонали, что визуально подтверждает наличие прямой связи между признаками.

3. Выборочный коэффициент корреляции вычислен двумя способами — по стандартной формуле и через условные варианты. Результаты совпали: $\bar{r}_{xy} = 0.8810$. Значение по негруппированным данным составило $r = 0.9101$.

4. Построены доверительные интервалы для $r_{xy}$: при $\gamma = 0.95$ — $(0.8305;\; 0.9171)$, при $\gamma = 0.99$ — $(0.8110;\; 0.9261)$. Оба интервала не содержат нуля и лежат в области сильной положительной корреляции.

5. Проверена гипотеза $H_0: r_{xy} = 0$ при $\alpha = 0.05$. Получено $|T_{\text{набл}}| = 19.26 > t_{\text{крит}} = 1.98$, гипотеза отвергнута. Корреляционная связь между признаками является статистически значимой.

Таким образом, между признаками $X$ и $Y$ существует сильная прямая линейная зависимость ($\bar{r}_{xy} \approx 0.88$), подтверждённая как доверительными интервалами, так и проверкой гипотезы о значимости.